hoodie-client/src/main/java/com/uber/hoodie/config/HoodieMemoryConfig.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2016 Uber Technologies, Inc. (hoodie-dev-group@uber.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *          http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.uber.hoodie.config;
+
+import javax.annotation.concurrent.Immutable;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.Properties;
+
+/**
+ * Memory related config
+ */
+@Immutable
+public class HoodieMemoryConfig extends DefaultHoodieConfig {
+
+  // This fraction is multiplied with the spark.memory.fraction to get a final fraction of heap space to use during merge
+  // This makes it easier to scale this value as one increases the spark.executor.memory
+  public static final String MAX_MEMORY_FRACTION_FOR_MERGE_PROP = "hoodie.merge.memory.fraction";
+  // Default max memory fraction during hash-merge, excess spills to disk
+  public static final String DEFAULT_MAX_MEMORY_FRACTION_FOR_MERGE = String.valueOf(0.6);
+  public static final String SPARK_EXECUTOR_MEMORY_PROP = "spark.executor.memory";
+  public static final String SPARK_EXECUTOR_MEMORY_FRACTION_PROP = "spark.memory.fraction";
+  public static final String DEFAULT_SPARK_EXECUTOR_MEMORY_FRACTION = "0.6";
```

Comment:
spark default is 0.5, are we bumping this up to 0.6? Can we get by without overriding spark default inside hudi ?

Reply:
It's not overriding. So there is a disconnect in what the spark docs say as default : https://spark.apache.org/docs/latest/configuration.html

and in the code : https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/SparkConf.scala#L507
I plan to open a ticket on Spark for this, for now, going with what is in the docs...

Reply:
Sure, still why are we defining spark configs defaults here? Spark config is something I totally want to leave it upto user while providing guidelines alone

Reply:
Just explained the reason for this in another comment, hope that suffices.

Reply:
Its orthogonal. I still don't understand why we are definiing spark config values in this file.. Can we for e.g decide dynamically based on values in `SparkConf`.
i.e `SPARK_EXECUTOR_MEMORY_PROP` , `SPARK_EXECUTOR_MEMORY_FRACTION_PROP` should not be defined in a HoodieConfig file right?  What am I missing?

Reply:
So, there are 2 ways to determine the spillable.memory.memory. We just let the user pass a fraction of available.memory that can be used to calculate this. In such a case, we want to get the `spark.executor.memory` and the `spark.memory.fraction`. To get these values from `SparkConf`, we need to know which properties to look for. Hence, these PROPs. The other option is to let the user figure out the spillable.memory and let the user correctly size the  spillable.memory when executory memory is increased or decreased, in which case we don't need the props or the calculation..

Reply:
> To get these values from SparkConf, we need to know which properties to look for 

is there constant inside Spark we can use without defining them here? thats the crux of what I am asking.. lets chat f2f, something is getting lost in translation

Reply:
I see it hardcoded everywhere in  the code, across tests and src's. Even in https://github.com/apache/spark/blob/5fa438471110afbf4e2174df449ac79e292501f8/core/src/main/scala/org/apache/spark/SparkConf.scala#L502 and https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/SparkContext.scala#L465

Reply:
The two constants are used in just the one method.. Can we move them as local final variables there.. I don't want to mix them with other hoodie level properties we declare here.. 

Also can we get away without defining the defaults here? i.e assume these values are passed down from app or spark installation?

Reply:
Move the variables as final. I think we cannot assume that these values are passed down from the app or spark installation since one can instantiate a HoodieWriteConfig even before a SparkContext or SparkConf is created. We could move this method to where it was earlier, in the HoodieIOHandle/ HoodieMergeHandle, then we can be sure that the SparkContext / SparkConf has been created. WDTY ? In the meantime, I can check more on this